### PR TITLE
Fix S2068 FN: add support for SecureString

### LINKED ISSUE #7323: Fix S2068 FN: add support for SecureString
## Why

We are working to improve the detection of issues in the Juliet benchmark. It heavily relies on `SecureString` for a lot of its tests.

[There currently are 284 FNs](https://sonarsource.atlassian.net/browse/APPSEC-731) that relate to hardcoded passwords (S2068) and use `SecureString.AppendChar(char c)`. Therefore, we want to make sure that methods using `SecureString` with a constant value are correctly detected.

## Detection Logic

An issue should be triggered if `SecureString.AppendChar(char c)` is called with a hardcoded `char`. Since this method only accepts one `char`, an issue should also be raised if `c` is one character of a hardcoded string instead (see Example.)

#### Sensitive

(This example is copied from the Juliet benchmark)

```csharp
private void Bla()
{
    using (SecureString securePwd = new SecureString())
    {
        for (int i = 0; i < "AP@ssw0rd".Length; i++)
        {
            securePwd.AppendChar("AP@ssw0rd"[i]); // Sensitive
        }

        // Do something with securePwd
    }
}
```

#### Compliant

```csharp
private void Bla(String password)
{
    using (SecureString securePwd = new SecureString())
    {
        for (int i = 0; i < password.Length; i++)
        {
            securePwd.AppendChar(password[i]); // Compliant
        }

        // Do something with securePwd
    }
}
```

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #7323

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/DoNotHardcodeCredentials.SecureString.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    class Tests
+    {
+        public void Characters()
+        {
+            using (var securePwd = new SecureString())
+            {
+                securePwd.AppendChar('P'); // Noncompliant
+                securePwd.AppendChar('a'); // Noncompliant
+                securePwd.AppendChar('s'); // Noncompliant
+                securePwd.AppendChar('s'); // Noncompliant
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/DoNotHardcodeCredentials.SecureString.cs:45]:
```suggestion
                securePwd.AppendChar('P'); // Noncompliant
                securePwd.AppendChar('a'); // Noncompliant
                securePwd.AppendChar('s'); // Noncompliant
                securePwd.AppendChar('s'); // Noncompliant
                var w = 'w';
                securePwd.AppendChar(w);   // FN
```
Adding this test will increase code coverage. Also, it needs to be fixed.

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/DoNotHardcodeCredentials.SecureString.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/DoNotHardcodeCredentials.SecureString.cs:45]: I'm not going to fix it. We already over-delivered and we said we don't want to invest in this deprecated API.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DoNotHardcodeCredentials.cs [hunk 2]
```diff
@@ -63,6 +63,19 @@ public sealed class DoNotHardcodeCredentials : DoNotHardcodeCredentialsBase<Synt
                         SyntaxKind.InvocationExpression);
                 });
 
+        protected override bool IsSecureStringAppendCharFromConstant(SyntaxNode argumentNode, SemanticModel model) =>
+            argumentNode is ArgumentSyntax { Expression: { } argumentExpression }
+            && argumentExpression switch
+            {
+                ElementAccessExpressionSyntax { Expression: { } accessed } => accessed.FindConstantValue(model) is string, // AppendChar("AP@ssw0rd"[i])
+                LiteralExpressionSyntax { RawKind: (int)SyntaxKind.CharacterLiteralExpression } => true, // AppendChar('P')
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DoNotHardcodeCredentials.cs:71]: Other literals like numbers or strings. It wouldn't compile with others, so it is more for correctness and being explicit about what we are looking for here.

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM

**Pull request merged**